pkg/pillar/cmd/downloader/resolveconfig.go

```diff
@@ -0,0 +1,194 @@
+	ctx.pubAppImgResolveStatus.Unpublish(config.Key())
+	log.Infof("handleAppImgResolveDelete for %s\n", key)
+}
+
+func publishResolveStatus(ctx *downloaderContext,
+	status *types.ResolveStatus) {
+
+	key := status.Key()
+	log.Debugf("publishResolveStatus(%s)\n", key)
+	pub := ctx.pubAppImgResolveStatus
+	pub.Publish(key, *status)
+	log.Debugf("publishResolveStatus(%s) Done\n", key)
+}
+
+func unpublishResolveStatus(ctx *downloaderContext,
+	status *types.ResolveStatus) {
+
+	key := status.Key()
+	log.Debugf("unpublishResolveStatus(%s)\n", key)
+	pub := ctx.pubAppImgResolveStatus
+	pub.Unpublish(key)
+	log.Debugf("unpublishResolveStatus(%s) Done\n", key)
+}
+
+func lookupResolveConfig(ctx *downloaderContext,
+	key string) *types.ResolveConfig {
+
+	sub := ctx.subAppImgResolveConfig
+	c, _ := sub.Get(key)
+	if c == nil {
+		log.Infof("lookupResolveConfig(%s) not found\n", key)
+		return nil
+	}
+	config := c.(types.ResolveConfig)
+	return &config
+}
+
+func lookupResolveStatus(ctx *downloaderContext,
+	key string) *types.ResolveStatus {
+
+	pub := ctx.pubAppImgResolveStatus
+	c, _ := pub.Get(key)
+	if c == nil {
+		log.Infof("lookupResolveStatus(%s) not found\n", key)
+		return nil
+	}
+	status := c.(types.ResolveStatus)
+	return &status
+}
+
+func resolveTagsToHash(ctx *downloaderContext, rc types.ResolveConfig) {
+	var (
+		err                           error
+		errStr, remoteName, serverURL string
+		syncOp                        zedUpload.SyncOpType = zedUpload.SyncOpGetObjectMetaData
+		trType                        zedUpload.SyncTransportType
+		auth                          *zedUpload.AuthInput
+	)
+
+	rs := lookupResolveStatus(ctx, rc.Key())
+	if rs == nil {
+		rs = &types.ResolveStatus{
+			AppID:        rc.AppID,
+			ImageID:      rc.ImageID,
+			Name:         rc.Name,
+			PurgeCounter: rc.PurgeCounter,
+		}
+	}
+
+	dst, errStr := lookupDatastoreConfig(ctx, rc.DatastoreID, rc.Name)
+	if errStr != "" {
+		rs.SetErrorInfo(agentName, errStr)
+		publishResolveStatus(ctx, rs)
+		return
+	}
+	// construct the datastore context
+	dsCtx, err := constructDatastoreContext(ctx, rc.Name, rc.NameIsURL, *dst)
+	if err != nil {
+		errStr := fmt.Sprintf("%s, Datastore construction failed, %s", rc.Name, err)
+		rs.SetErrorInfo(agentName, errStr)
+		publishResolveStatus(ctx, rs)
+		return
+	}
+
+	log.Infof("Resolving config <%s> using %v allow non-free port\n",
+		rc.Name, rc.AllowNonFreePort)
+
+	var addrCount int
+	if !rc.AllowNonFreePort {
+		addrCount = types.CountLocalAddrFreeNoLinkLocal(ctx.deviceNetworkStatus)
+		log.Infof("Have %d free management port addresses\n", addrCount)
+		err = errors.New("No free IP management port addresses for download")
+	} else {
+		addrCount = types.CountLocalAddrAnyNoLinkLocal(ctx.deviceNetworkStatus)
+		log.Infof("Have %d any management port addresses\n", addrCount)
+		err = errors.New("No IP management port addresses for download")
+	}
+	if addrCount == 0 {
+		errStr = err.Error()
+	}
```

Comment:
Shouldn't you return the error here?

Reply:
Thanks for pointing it out. Returning error now.